ci: Remove deprecated logic from the ci module (#47062)
LLNL/spack: lib/spack/docs/pipelines.rst

------------------

The simplest fully functional standalone example of a working pipeline can be
examined live at this example `project <https://gitlab.com/spack/pipeline-quickstart>`_
on gitlab.com.

Here's the ``.gitlab-ci.yml`` file from that example that builds and runs the
pipeline:

.. code-block:: yaml

   stages: [ "generate", "build" ]

   variables:
     SPACK_REPOSITORY: "https://github.com/spack/spack.git"

>      SPACK_REF: "develop-2024-10-06"

Reviewer: Why not use the `develop` ref here?

Author: In general I don't think it's a good idea to run pipelines with a branch that has the potential to change as frequently as `develop`. In my opinion, the probability of running `generate` with one sha and subsequent build jobs with a different one should be zero.

Reviewer: That's fair. My worry is that we are likely going to start dropping snapshot tags before long and only keep the last 6 and this tag won't exist anymore. I don't want to advocate for a hash, I don't think that is better.

Since we haven't started pruning snapshots maybe we can just keep this in mind for things we need to be aware of.

Author: I was thinking once we have a release with your mirror credentials change in it, we could switch to that released version.

   spack:
     view: false
     concretizer:
       unify: true
       reuse: false

     definitions:
     - pkgs:
       - zlib
       - bzip2 ~debug
     - compiler:
       - '%gcc'

     specs:
     - matrix:
       - - $pkgs
       - - $compiler

     ci:
       target: gitlab

       pipeline-gen:
       - any-job:
           tags:
             - saas-linux-small-amd64
           image:
             name: ghcr.io/spack/ubuntu20.04-runner-x86_64:2023-01-01
           before_script:
           - git clone ${SPACK_REPOSITORY}
           - cd spack && git checkout ${SPACK_REF} && cd ../
           - . "./spack/share/spack/setup-env.sh"
           - spack --version
           - export SPACK_USER_CONFIG_PATH=${CI_PROJECT_DIR}
           - spack config blame mirrors

.. note::
   The use of ``reuse: false`` in spack environments used for pipelines is
   almost always what you want, as without it your pipelines will not rebuild

>    packages whenever the hashes change.

Reviewer: ```suggestion
   packages even if the package recipe has changed. This is due to concretizer strongly preferring known hashes over computing new hash when ``reuse: true``.
```

The ``ci`` section in the above environment file contains the bare minimum

> configuration required for ``spack ci genrate`` to create a working pipeline.

Reviewer 2: ```suggestion
configuration required for ``spack ci generate`` to create a working pipeline.
```

The ``target: gitlab`` tells spack that the desired pipeline output is for
gitlab.  However, this isn't strictly required, as currently gitlab is the
only possible output format for pipelines. The ``pipeline-gen`` section
contains the key information needed to specify attributes for the generated
jobs.  Notice that it contains a list which has only a single element in
this case.  In real pipelines it will almost certainly have more elements,
and in those cases, order is important: spack starts at the bottom of the
list and works upwards when applying attributes.

But in this simple case, we use only the special key ``any-job`` to
indicate that spack should apply the specified attributes (``tags``, ``image``,
and ``before_script``) to any job it generates.  This includes jobs for
building/pushing all packages, a ``rebuild-index`` job at the end of the
pipeline, as well as any ``noop`` jobs that might be needed by gitlab when
no rebuilds are required.

Something to note is that in this simple case, we rely on spack to

> generate a reasonble script for the package build jobs (it just creates

Reviewer 2: ```suggestion
generate a reasonable script for the package build jobs (it just creates
```

a script that invokes ``spack ci rebuild``).

Another thing to note is the use of the ``SPACK_USER_CONFIG_DIR`` environment
variable in any generated jobs.  The purpose of this is to make spack
aware of one final file in the example, the one that contains the mirror
configuration.  This file, ``mirrors.yaml`` looks like this:

.. code-block:: yaml

   mirrors:
     buildcache-destination:

>        url: oci://registry.gitlab.com/spack/pipeline-quickstart

Reviewer: I think this should probably be `oci://${CI_REGISTRY_IMAGE}` to be portable across gitlab instances. The requirement for this variable is the registry must be enabled for the project, we is already required to run the example as self contained.

The URL field in `mirrors.yaml` supports variable expansion.

Author: That makes sense in theory, but if that were working, I don't think we would have needed to have a dedicated [step](https://github.com/spack/spack/blob/develop/share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml#L138-L140) in our generate script to expand variables in mirror configs.  Or am I forgetting some other reason we have to do it that way?

Reviewer: It may not have been added at the time? I dug into the code and it _should_ work.

Author: Maybe we can keep it like this for now, and assuming it works, follow up with that small change when we're in there changing the `SPACK_REF`.

       binary: true
       access_pair:
         id_variable: CI_REGISTRY_USER
         secret_variable: CI_REGISTRY_PASSWORD

Note the name of the mirror is ``buildcache-destination``, which is required
as of Spack 0.23 (see below for more information).  The mirror url simply

> points to the container registry associated with the project, and the file

Reviewer: ```suggestion
points to the container registry associated with the project, and the ``id_variable`` and ``secret_variable``
```
